### ISSUE TITLE
Poor optimization of IN and OR with ORC adapter

### ISSUE THREAD

**Issue by Downchuck** [facebook/presto#3023]
This one came up while working with S3 remotely:
Values have XXXs though the data is publicly available. They are ints.

Running the three values independently:

```
presto:ssdm> select * from s3_ssdm where ssn = 297XXXXX;
(1 row)
0:04 [10K rows, 193KB] [2.53K rows/s, 48.7KB/s]

presto:ssdm> select * from s3_ssdm where ssn = 508XXXXX;
(1 row)
0:03 [10K rows, 188KB] [3.86K rows/s, 72.6KB/s]

presto:ssdm> select * from s3_ssdm where ssn = 108XXXXX;
(1 row)
0:03 [10K rows, 194KB] [3.87K rows/s, 75.2KB/s]
```

Running the three values in an IN query:

```
presto:ssdm> select * from s3_ssdm where ssn in (108XXXXX, 508XXXXX, 297XXXXX);
(3 rows)
0:48 [61.4M rows, 583MB] [1.28M rows/s, 12.1MB/s]
```

Note that nearly the entire file is scanned when using the in query.

**Comment by Downchuck** [comment]
Simplified case:

Bad:
```
select * from s3_ssdm where ssn in (SELECT 108XXXXX);
(1 row)
1:07 [87.7M rows, 840MB] [1.31M rows/s, 12.5MB/s]
```

Good:
```
presto:ssdm> select * from s3_ssdm where ssn = 108XXXXX;
(1 row)
0:03 [10K rows, 194KB] [3.24K rows/s, 62.9KB/s]
```

**Comment by Downchuck** [comment]
Workaround:
```
select * from s3_ssdm where ssn = 108XXXXX
union
select * from s3_ssdm where ssn = 508XXXXX
union
select * from s3_ssdm where ssn = 297XXXXX
```

**Comment by nezihyigitbasi** [comment]
Not sure this is only an issue with the orc adapter. We are using parquet and I have also observed some  performance degradation with `in` vs. `=`. Basically in my case when I unrolled the `in` part into a series of `=` checks the performance increased noticeably.

**Comment by Downchuck** [comment]
@nezihyigitbasi OR checks are falling apart in ORC as well. Unrolled to union to workaround.

**Comment by snarayananqubole** [comment]
@Downchuck you might want to see this thread https://groups.google.com/forum/#!searchin/presto-users/tupledomain/presto-users/LgeSPdUXztU/J6LPHTROyIkJ

**Comment by Downchuck** [comment]
@snarayananqubole Thank you very much.

I'm seeing this one: "As a short term fix, we could be smarter about the compaction, and only compact large domain objects."

And I'm thinking, as a near-term fix, we can have a session setting:
https://prestodb.io/docs/current/sql/set-session.html

Something like hive.disable_compact_domains = true;
That'd be a big help, in short time.

-Charles

**Comment by Downchuck** [comment]
Well, harder than it looked. I spent some time but could not figure out how to get the session object over to the SplitManager. So given that, I will just use the other hack, of checking length.

Here's where I got on trying to set a session property.

```
diff --git a/presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java b/presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java
index d86d64b..b0cdbef 100644
--- a/presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java
+++ b/presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java
@@ -28,6 +28,7 @@ public final class HiveSessionProperties
     private static final String ORC_MAX_MERGE_DISTANCE = "orc_max_merge_distance";
     private static final String ORC_MAX_BUFFER_SIZE = "orc_max_buffer_size";
     private static final String ORC_STREAM_BUFFER_SIZE = "orc_stream_buffer_size";
+    private static final String DISABLE_COMPACT_DOMAINS = "disable_compact_domains";

     private HiveSessionProperties()
     {
@@ -48,6 +49,10 @@ public final class HiveSessionProperties
         }
     }

+    public static boolean isDisableCompactDomains(ConnectorSession session, boolean defaultValue) {
+        return isEnabled(DISABLE_COMPACT_DOMAINS, session, defaultValue);
+    }
+
     public static boolean isOptimizedReaderEnabled(ConnectorSession session, boolean defaultValue)
     {
         return isEnabled(OPTIMIZED_READER_ENABLED, session, defaultValue);
diff --git a/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java b/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java
index cae5f12..371cb02 100644
--- a/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java
+++ b/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java
@@ -88,6 +88,9 @@ import static java.lang.String.format;
 import static org.apache.hadoop.hive.metastore.ProtectMode.getProtectModeFromString;
 import static org.apache.hadoop.hive.metastore.Warehouse.makePartName;

+import static com.facebook.presto.hive.HiveSessionProperties.isOptimizedReaderEnabled;
+import com.facebook.presto.spi.ConnectorSession;
+
 public class HiveSplitManager
         implements ConnectorSplitManager
 {
@@ -112,6 +115,7 @@ public class HiveSplitManager
     private final boolean forceLocalScheduling;
     private final boolean recursiveDfsWalkerEnabled;
     private final boolean assumeCanonicalPartitionKeys;
+    private final ConnectorSession session;

     @Inject
     public HiveSplitManager(
@@ -176,6 +180,7 @@ public class HiveSplitManager
         this.forceLocalScheduling = forceLocalScheduling;
         this.recursiveDfsWalkerEnabled = recursiveDfsWalkerEnabled;
         this.assumeCanonicalPartitionKeys = assumeCanonicalPartitionKeys;
+	this.session = handle.getConnectorSession();;
     }

     @Override
@@ -218,6 +223,10 @@ public class HiveSplitManager

     private static TupleDomain<HiveColumnHandle> toCompactTupleDomain(TupleDomain<ColumnHandle> effectivePredicate)
     {
+	if(isDisableCompactDomains(session, false)) {
+		return effectivePredicate;
+	}
+
         ImmutableMap.Builder<HiveColumnHandle, Domain> builder = ImmutableMap.builder();
         for (Map.Entry<ColumnHandle, Domain> entry : effectivePredicate.getDomains().entrySet()) {
             HiveColumnHandle hiveColumnHandle = checkType(entry.getKey(), HiveColumnHandle.class, "ConnectorColumnHandle");
```

**Comment by dain** [comment]
Compact domains was put in place because we were seeing multi megabyte domains after the predicate push down optimizer runs (this optimizer combines predicates with extra information like which partitions are available).  When this happens, all request start to fail as the domains become bigger than the buffers allow.  A solution to this problem would be to estimate the size of the domain (say count the nodes in the tree), and only compact when it exceeds some threshold (say 1k nodes).

**Comment by Downchuck** [comment]
@dain That is a good plan... Does that simply mean running an if statement checking count in that toCompactDomain method?

**Comment by dain** [comment]
That is what I was thinking

**Comment by Downchuck** [comment]
My understanding of the data structures on this project is still weak.

Are we inside of the range loop, or is it even simpler, effectivePredicate.getDomains().length() kind of check?


```
diff --git a/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java b/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java
index cae5f12..9d9e9e5 100644
--- a/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java
+++ b/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java
@@ -223,9 +223,11 @@ public class HiveSplitManager
             HiveColumnHandle hiveColumnHandle = checkType(entry.getKey(), HiveColumnHandle.class, "ConnectorColumnHandle");

             SortedRangeSet ranges = entry.getValue().getRanges();
-            if (!ranges.isNone()) {
-                // compact the range to a single span
-                ranges = SortedRangeSet.of(entry.getValue().getRanges().getSpan());
+            if (ranges.getRangeCount() > 1000) {
+                    if (!ranges.isNone()) {
+                        // compact the range to a single span
+                        ranges = SortedRangeSet.of(entry.getValue().getRanges().getSpan());
+                    }
             }

             builder.put(hiveColumnHandle, new Domain(ranges, entry.getValue().isNullAllowed()));
```

**Comment by snarayananqubole** [comment]
Yup - that's the hack we've used in our internal deployment. We've tested out up to 100 values in the IN list and seems to work well.

**Comment by Downchuck** [comment]
Do you run a value sort?
Seems like good hygiene.

**Comment by Downchuck** [comment]
Is there anything additional that needs to be done on this? I'd like to get it into an upcoming Presto release.

**Comment by Downchuck** [comment]
@dain How can I get this patch applied? I don't have CLA authority on this end, it's a lot to ask of our attorney for a basic patch.

**Comment by Downchuck** [comment]
@electrum saw you were working in some ORC code. Any chance you could apply this fix?

**Comment by electrum** [comment]
Fixed in #3326. @Downchuck I added a config property to this which defaults to `100`
